### PR TITLE
feat: Parse go.mod

### LINKED ISSUE #12: parser: create code to parse the go.mod
The go.mod contains the dependencies with the following values:

* module name (string)
* go version (string)
* Array of Dependencies (maybe split between direct and indirect?)
   * URL (string): showing where the dependency is
   * Version (string): which version of the dependency
   * Indirect (bool): whether the dependency is indirect or not

See https://pkg.go.dev/cmd/go#hdr-The_go_mod_file for more details.

## URL

This ideally is public and available on github.com. However, it could also be on cloud.google.com, code.cloudfoundry.org, go.opentelemetry.io, golang.org, gopkg.in, k8s.io, modernc.org, sigs.k8s.io, etc.

## Version

Ideally a tag or possible release. If not it could be date + hash (e.g. `v0.0.0-20220728103510-ee6ede2d64ed`)

## Indirect

Will be followed by `// indirect`

## Parsing

Expect a byte[] and then return go.mod struct with the above data.
May consider looking at how upstream handles the `go mod *` functions when parsing.

### PR DESCRIPTION
**Pull request by sspaink**
resolve: https://github.com/goreapers/gograveyard/issues/12

Added a function `Parse` to take a byte array from a `go.mod` file and parse it into a struct that contains slices of all the modules. The module is defined by the path, version and categorized if it is indirect or not.

### CODE REVIEW

File: modfile.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+			continue
+		}
+
+		// If in a require section, parse the module
+		// A module is required to have a path and version, the "indirect" comment is optional
+		m := strings.Split(strings.TrimSpace(scanner.Text()), " ")
+		const pathAndVersion = 2
+		// Skip if invalid module, missing required path or version
+		if len(m) < pathAndVersion {
+			continue
```
Comment (powersj) [inline modfile.go:59]: Is it worth putting a "warning" here for now to see what we may end up skipping?
Is there a test case we can add as well?

File: modfile.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+	var require bool
+	var f ModFile
+
+	for scanner.Scan() {
+		switch scanner.Text() {
+		case "require (":
+			require = true
+
+			continue
```
Comment (powersj) [inline modfile.go]:
```suggestion
			require = true
			continue
```

File: modfile.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+		case ")":
+			require = false
+
+			continue
```
Comment (powersj) [inline modfile.go]:
```suggestion
			require = false
			continue
```

**Review by powersj — COMMENTED** [review]
Thank you for doing this! Looks great, couple questions in line.

Additionally, when looking at the upstream parser, they had cases for [replace](https://golang.org/ref/mod#go-mod-file-replace), [retract](https://golang.org/ref/mod#go-mod-file-retract), [exclude](https://golang.org/ref/mod#go-mod-file-exclude). Do we know what happens with those?

File: modfile.go (hunk 2 above)
Comment (sspaink) [reply modfile.go]: `continue with no blank line before (nlreturn)go-golangci-lint` I get this linter error complaining about the missing space, should I disable it perhaps?

Comment (powersj) [reply modfile.go]: oh interesting! yeah I'm cool with disabling that one as it seems a little weird

**Comment by sspaink** [comment]
@powersj Good find regarding replace, retract, and exclude. So I pushed a change to handle `replace` directives, let me know what you think! I am not sure what to do regarding retract and exclude, they seem to the guards to prevent people from using certain versions and packages? I don't think it changes the dependencies declared in the mod file?

**Review by powersj — APPROVED** [review]
This looks great!

> I am not sure what to do regarding retract and exclude, they seem to the guards to prevent people from using certain versions and packages? 

correct, just something we will want to not barf on

> I don't think it changes the dependencies declared in the mod file?

Right

**Pull request merged**